src/System.Collections/tests/Generic/Comparers/Comparer.Generic.Tests.cs

```diff
@@ -0,0 +1,137 @@
+                Assert.True(cloned.Equals(comparer));
+                Assert.True(comparer.Equals(cloned));
+
+                // Equals() should not return true for null
+                // Prevent a faulty implementation like Equals(obj) => obj is FooComparer<T>, which will be true for null
+                Assert.False(cloned.Equals(null));
+            }
+        }
+
+        [Fact]
+        public void GetHashCodeShouldBeOverriddenAndBeTheSameAsLongAsTheTypeIsTheSame()
+        {
+            var comparer = Comparer<T>.Default;
+
+            // Multiple invocations should return the same result,
+            // whether GetHashCode() was overridden or not
+            Assert.Equal(comparer.GetHashCode(), comparer.GetHashCode());
+
+            // If we are running on full framework/CoreCLR, Comparer<T> additionally
+            // overrides the Equals(object) and GetHashCode() methods for itself,
+            // presumably to support serialization, so test that behavior as well.
+            // This is not done in .NET Native.
+            if (!RuntimeDetection.IsNetNative)
+            {
+                var cloned = ObjectCloner.MemberwiseClone(comparer);
+                Assert.Equal(cloned.GetHashCode(), cloned.GetHashCode());
+
+                // Since comparer and cloned should have the same type, they should have the same hash
+                Assert.Equal(comparer.GetHashCode(), cloned.GetHashCode());
+            }
+        }
+
+        [Fact]
+        public void IComparerCompareWithObjectsNotOfMatchingTypeShouldThrow()
+        {
+            // Comparer<T> implements IComparer for back-compat reasons.
+            // The explicit implementation of IComparer.Compare(object, object) should
+            // throw if both inputs are non-null and one of them is not of type T
+            IComparer comparer = Comparer<T>.Default;
+            Task<T> notOfTypeT = Task.FromResult(default(T));
+            if (default(T) != null) // if default(T) is null these asserts will fail as IComparer.Compare returns early if either side is null
+            {
+                Assert.Throws<ArgumentException>(() => comparer.Compare(notOfTypeT, default(T))); // lhs is the problem
+                Assert.Throws<ArgumentException>(() => comparer.Compare(default(T), notOfTypeT)); // rhs is the problem
+            }
+            else if (!typeof(T).GetTypeInfo().IsAssignableFrom(typeof(Task<T>))) // catch cases where Task<T> actually is a T, like object or non-generic Task
+            {
+                Assert.Throws<ArgumentException>(() => comparer.Compare(notOfTypeT, notOfTypeT)); // The implementation should not attempt to short-circuit if both sides have reference equality
+            }
+            Assert.Throws<ArgumentException>(() => comparer.Compare(notOfTypeT, Task.FromResult(default(T)))); // And it should also work when they don't
```

Comment:
For all of these and other argument exception checks, can we add parameter name validation as well via the first argument to Throws?